src/Constraint/Constraint.php

```diff
@@ -128,13 +151,18 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
     }
 
     /**
-     * @param Constraint $provider
+     * @param ConstraintInterface $provider
      * @param bool $compareBranches
      *
      * @return bool
      */
-    public function matchSpecific(Constraint $provider, $compareBranches = false)
```

Comment:
except that we should aim at removing the child classes from Composer itself in the future (they are there for BC reasons). Having a package here and using child classes in composer looks weird. It should end up using the package rather than extending classes